### PR TITLE
Implement retain as publish subscribe flag

### PR DESCRIPTION
**Pull request by larshesel**
This is the first part of implementing the subscription options and this implements the Retain as Publish flag.

Currently working on implementing the 'no local' and 'retain handling' flags, but I though it would be nice to review this first as it's already quite large as is and touches a lot of different places in the code base.

### CODE REVIEW

File: apps/vmq_server/src/vmq_reg.erl [hunk 1]
```diff
@@ -597,12 +619,21 @@ fold_subscribers(FoldFun, Acc) ->
               FoldFun(SubscriberId, Subs, AccAcc)
       end, Acc).
 
--spec add_subscriber([{topic(), qos() | not_allowed}], subscriber_id()) -> ok.
+-spec add_subscriber([{topic(), qos() | not_allowed} |
+                      {topic(), qos() | not_allowed, map()}], subscriber_id()) -> ok.
```
Comment (dergraf) [inline apps/vmq_server/src/vmq_reg.erl]: Shouldn't the spec looks like: 
```erlang
-spec add_subscriber([{topic(), qos() | not_allowed} |
                      {topic(), {qos() | not_allowed, map()}}], subscriber_id()) -> ok.
```

File: apps/vmq_server/src/vmq_subscriber.erl [hunk 2]
```diff
@@ -173,13 +173,19 @@ get_node_subs(Node, Subs) ->
 -spec fold(function(), any(), subs() | changes()) -> any().
 fold(Fun, Acc, Subs) ->
     lists:foldl(fun({Node, _, NSubs}, AccAcc) ->
-                        lists:foldl(fun({Topic, QoS}, AccAccAcc) ->
-                                            Fun({Topic, QoS, Node}, AccAccAcc)
-                                    end, AccAcc, NSubs);
+                        lists:foldl(
+                          fun({Topic, QoS}, AccAccAcc) ->
+                                  Fun({Topic, QoS, Node}, AccAccAcc);
+                             ({Topic, QoS, Opts}, AccAccAcc) ->
```
Comment (dergraf) [inline apps/vmq_server/src/vmq_subscriber.erl]: Rename QoS to SubInfo, and remove never matching case on linen 179

File: apps/vmq_server/src/vmq_subscriber.erl [hunk 3]
```diff
@@ -173,13 +173,19 @@ get_node_subs(Node, Subs) ->
...
+                                  Fun({Topic, {QoS, Opts}, Node}, AccAccAcc)
+                          end, AccAcc, NSubs);
                    ({Node, NChanges}, AccAcc) ->
-                        lists:foldl(fun({Topic, QoS}, AccAccAcc) ->
-                                            Fun({Topic, QoS, Node}, AccAccAcc)
-                                    end, AccAcc, NChanges)
+                        lists:foldl(
+                          fun({Topic, QoS}, AccAccAcc) ->
+                                  Fun({Topic, QoS, Node}, AccAccAcc);
+                             ({Topic, QoS, Opts}, AccAccAcc) ->
```
Comment (dergraf) [inline apps/vmq_server/src/vmq_subscriber.erl]: See Comment above

File: apps/vmq_server/src/vmq_reg.erl (hunk 1 above)
Comment (larshesel) [reply apps/vmq_server/src/vmq_reg.erl]: Fixed

File: apps/vmq_server/src/vmq_subscriber.erl (hunk 2 above)
Comment (larshesel) [reply apps/vmq_server/src/vmq_subscriber.erl]: Fixed

File: apps/vmq_server/src/vmq_subscriber.erl (hunk 3 above)
Comment (larshesel) [reply apps/vmq_server/src/vmq_subscriber.erl]: Fixed

**Pull request merged**